### ISSUE TITLE
[CI Failure] vLLM nightly - [WH-GLX] Llama-3.3-70B-Instruct device with sampling-tests (prefix caching) - Intermittent failure

### ISSUE THREAD

**Issue by tchedaTT** [tenstorrent/tt-metal#48635]
### Pipeline Name

vLLM nightly tests

### Test Name

[WH-GLX] Llama-3.3-70B-Instruct device with sampling-tests (prefix caching)

### Path to Test in Repo

.github/workflows/vllm-nightly-tests-impl.yaml (multiple commands)

### Non-Deterministic (ND) failure?

Yes

### Failure type

Hang

### Link to Failing Pipeline for non-ND or Custom Dispatch Run for ND

https://github.com/tenstorrent/tt-metal/actions/runs/28139255072/job/83335013970#step:19:6945

### Failure Signature Snippet

```shell
2026-06-25 01:32:23.116 | error    |           Metal | Timeout detected (metal_context.cpp:771)
```

### Additional info

```shell
I scanned scheduled `vllm-nightly-tests.yaml` runs from 2026-04-01 through 2026-06-29. Earlier runs are no longer available. Last fully green vllm nightly was early february.


- Matching scheduled jobs scanned: 89
- Timeout-like hangs found: 17 / 89, about 19%
- Non-cancelled jobs: 85, so 17 / 85, about 20%
- Full job successes in that same window: 7 / 89
- Separate sampler concat fatals excluded from this count: 2, on 2026-06-26 and 2026-06-28

This is intermittent: the same job has passed on `g04glx01`, `g04glx03`, `g05glx04`, `g08glx04`, `OM1-01A02-STGWH01`, and `OM1-01A02-STGWH03`.

### Timeline

Earliest confirmed timeout in scanned history:

- 2026-04-14, run #1529, job 71185210907, runner `g04glx01`
- https://github.com/tenstorrent/tt-metal/actions/runs/24374330735/job/71185210907
- Stage: sampling test `test_min_tokens`; server reached prefill trace / `TtLlamaPrefetcherSetup`
- Error: `Metal | Timeout detected`

Latest scheduled-nightly timeout found:

- 2026-06-25, run #1952, job 83335013970, runner `OM1-01A02-STGWH02`
- https://github.com/tenstorrent/tt-metal/actions/runs/28139255072/job/83335013970
- Stage: traced prefill replay, `prompt_lens=2048`, `prefill_seq_len=2048`, last seen around `Prefilling User 14`
- Triage running ops included `ReduceScatterMinimalAsyncDeviceOperation` and `AllGatherAsyncDeviceOperation`

Passing scheduled runs for that job since April 1, 2026:
2026-04-21
2026-04-24
2026-05-03
2026-05-04
2026-05-17
2026-06-08
2026-06-19

### Runners Where Timeout Was Seen

From scheduled runs only, 2026-04-01 through 2026-06-29:

| Runner | Timeout-like occurrences |
|---|---:|
| `g04glx01` | 8 |
| `g04glx03` | 2 |
| `g05glx04` | 2 |
| `OM1-01A02-STGWH02` | 2 |
| `OM1-01A01-STGWH03` | 1 |
| `OM1-01A03-STGWH03` | 1 |
| `OM1-01A04-STGWH02` | 1 |

### Observed Hang Stages

The hang does not have one single fixed op/stage. It has appeared in at least these places:

- `TtLlamaPrefetcherSetup`
- decode warmup after `Starting decode warmup` / `Warming up decode`
- traced prefill replay with `prefill_seq_len=1024`
- traced prefill replay with `prefill_seq_len=2048`
- prefill sequence length 128 in the PR reproduction

Triage running ops varied across occurrences:

- `TilizeDeviceOperation`
- `UntilizeDeviceOperation`
- `TypecastDeviceOperation`
- `SDPAOperation`
- `EmbeddingsDeviceOperation`
- `AllGatherAsyncDeviceOperation`
- `ReduceScatterMinimalAsyncDeviceOperation`

# Timeout ablation
I reran the nightly with TT_METAL_OPERATION_TIMEOUT_SECONDS=60 and the test still hung, so this is not just a case of too tight timeouts.
```

**Comment by tchedaTT** [comment]
cc @viktorpusTT @mtairum @djordje-tt @sraizada-tt

**Comment by tchedaTT** [comment]
Interestingly this doesn't happen in tt-inference-server, or at least not consistently.
The test also doesn't hang eg. here https://github.com/tenstorrent/tt-metal/actions/runs/28393646068 and here https://github.com/tenstorrent/tt-metal/actions/runs/27275460979/job/80557797302

**Comment by tchedaTT** [comment]
A similar timeout happens in e2e tests for qwen which uses the same codebase, so this is not vllm specific https://github.com/tenstorrent/tt-metal/actions/runs/28417909759/job/84206485146#step:6:911

**Comment by tchedaTT** [comment]
@yieldthought autodebug

**Comment by viktorpusTT** [comment]
This is not specific to this model. Metal timeouts appear also in other GLX tests, in different phases (before warmup, during warmup, during benchmark). Examples:
* https://github.com/tenstorrent/tt-metal/actions/runs/28412468892/job/84195679376#step:19:2646
* https://github.com/tenstorrent/tt-metal/actions/runs/28209887137/job/83570685649#step:19:439

**Comment by yieldthought** [comment]
✅ AutoDebug top finding: Galaxy decode may reuse a stale DRAM prefetcher address table, causing wrong memory reads and timeouts.

<details>
<summary>Full report</summary>

# AutoDebug Report: tt-metal issue

Inspection-only investigation. I did not run the vLLM repro or any hardware-dependent test.

## Headline findings

### 1. Galaxy decode can run `ttnn.dram_prefetcher` with a stale or wrong address table

The most timeout-capable bug I found is in the Galaxy DRAM prefetcher setup. `TtLlamaPrefetcherSetup` stores the device tensor that contains weight buffer addresses in a module-global variable:

- `models/demos/llama3_70b_galaxy/tt/prefetcher_common.py:13` defines `global_tt_tensor_address`.
- `get_input_tensors()` only rebuilds it when the global is `None` (`prefetcher_common.py:146-153`).
- The cached value is not keyed by mesh device, submesh, model instance, number of tensors, layer count, or the actual buffer addresses.
- The model destructor tries to reset `global_tt_tensor_address`, but it only imported `TtLlamaPrefetcherSetup`, not the module global. The `global global_tt_tensor_address` in `llama_model.py:883-885` therefore resets a `llama_model.py` name, not `prefetcher_common.global_tt_tensor_address`.
- The regular `text_demo.py` has a manual module-level reset with a comment saying this avoids reusing the previous test's address tensor and causing a device mismatch (`text_demo.py:839-843`). The vLLM/server path should not rely on test-only cleanup.

The lowered C++ path makes this dangerous. `dram_prefetcher` treats the last input tensor as the address tensor (`dram_prefetcher_program_factory.cpp:48-50`) and binds its buffer as the address CB (`dram_prefetcher_program_factory.cpp:169-178`). The reader kernel then directly loads `tensor_addrs_l1[layer * num_tensors + t]` and uses that value as the DRAM base address for NOC reads (`reader_dram.cpp:37-49`). Validation checks that the address tensor is on the same device and has the right layout/type, but it does not check that the address table contents match the current input tensors (`dram_prefetcher_device_operation.cpp:70-80`).

That gives a complete timeout chain:

1. A model/test/server lifecycle builds or reuses `global_tt_tensor_address` for one prefetcher setup.
2. A later decode setup inserts the current layer weight tensors, but `get_input_tensors()` reuses the old address tensor instead of rebuilding from `self.tensor_addrs`.
3. `ttnn.dram_prefetcher` receives current tensors plus stale buffer addresses.
4. The prefetcher reader DM kernels read the wrong DRAM locations and feed the global CB with wrong or missing weight blocks.
5. Downstream matmul/CCL/trace work can stall later; the Metal timeout is then reported at a sync/readback/teardown boundary, not necessarily at the bad prefetch launch.

This fits the observed `TtLlamaPrefetcherSetup` and decode-warmup hangs, and it fits the Qwen comment because Qwen uses the same Galaxy prefetcher code. It does not by itself explain GLX tests that do not use this Galaxy prefetcher path.

Suggested fix direction: remove the module global, or key it by `(mesh_device/submesh identity, n_layers, n_tensors, tensor buffer address tuple)` and rebuild on any mismatch. Reset it by importing `prefetcher_common` and assigning `prefetcher_common.global_tt_tensor_address = None`, not through a local `global` in `llama_model.py`.

### 2. Galaxy batch-32 prefill warmup builds malformed page-table rows

The prefill warmup path creates `page_table = torch.zeros(1, 1)` (`generator.py:1903-1907`) and then sweeps batch sizes `(1, 32)` (`generator.py:327-350`). For batch 32 it pads the missing 31 rows with `-1`, but only for the single existing column (`generator.py:341-346`).

Later `_get_prefill_user_page_table()` expands that to the number of blocks required by the warmup length. Missing columns are padded with `0`, then all 32 rows are treated as active batched users (`generator.py:1878-1899`). For any warmup length requiring more than one KV block, rows 1-31 become `[-1, 0, 0, ...]`.

That table is flattened for batched prefill in `prepare_prefill_inputs_host()` (`llama_model.py:270-290`) and consumed by `paged_fill_cache()` in attention (`llama_attention.py:715-731`). The paged-fill writer kernel treats `-1` as "skip" but treats `0` as physical block 0 (`writer_fill_cache_interleaved.cpp:12-32`, `181-199`). So inactive warmup slots skip their first block and then write subsequent blocks into physical block 0.

This is a clear source bug and can poison the KV/null block that prefix-cached paths deliberately use as read-safe padding. I am less confident that this alone causes the Metal timeout rather than bad output or later undefined behavior, so I would treat it as a serious contributing bug/test gap unless hardware logs show the hang follows this warmup write.

Suggested fix direction: for warmup, allocate a page table with `batch x num_blocks` before padding rows, and fill inactive batched rows with `-1` for all columns. Alternatively skip batched paged-fill warmup unless it is using a real vLLM-style page table.

## Supporting evidence: the timeout line is a symptom

`Metal | Timeout detected (metal_context.cpp:771)` is emitted by `MetalContext::on_dispatch_timeout_detected()` after dispatch/completion progress has already stopped. The main wait paths are completion queue waits and core-done waits, not ordinary op enqueue. Galaxy trace replay is explicitly nonblocking (`generator.py:1355`, `generator.py:1615`), so the visible timeout can appear later than the operation that produced the bad state.

This matches the issue matrix: triage saw many different "running ops" (`AUTODEBUG_TICKET.md:92-110`) and the 60s timeout ablation still hung (`AUTODEBUG_TICKET.md:112-113`).

## Other potential issues and adjudicated non-findings

- Galaxy trace input refresh uses `copy_host_to_device()` without an explicit `cq_id`, then replays the trace with `cq_id=0` (`generator.py:1343-1355`). If an async caller has pushed a nonzero current command queue, the H2D input update and trace replay are unordered. I did not find an in-repo Galaxy path that pushes a nonzero CQ, so this needs vLLM/plugin confirmation before headlining.
- Llama, but not Qwen, has a double `setup_prefill()` in `TtTransformer.__init__()` (`llama_model.py:128-134`). Layer-level CCL references can differ from `self.model.tt_ccl`; the generator even comments on this (`generator.py:281-285`). This is a real state-ownership smell, but I could not complete a causal chain to the reported Qwen-like timeout.
- I initially suspected `reduce_scatter_minimal_async` cache-hit runtime-argument override because it copies the `GetRuntimeArgs()` return value. After review, that is not a no-op: `RuntimeArgsData` is a pointer wrapper (`runtime_args_data.hpp:18-24`), so copying the vectors still leaves entries pointing to live runtime-argument storage. I do not consider that a headline root cause.
- Broad program-cache or async-scheduling races remain possible, but I did not find enough in-repo evidence that the named Galaxy vLLM path concurrently enters one model instance or one program-cache entry.

## Why the headline is scoped this way

The named failure is a Galaxy Llama/Qwen code path, and both concrete bugs above are in that path. The prefetcher address-table bug is the strongest timeout explanation because it can make device kernels read arbitrary DRAM addresses and stall downstream work. The warmup page-table bug is definitely wrong and hits the prefix-caching warmup surface, but the static evidence most directly proves KV-cache corruption, not a deterministic deadlock.

I would fix the prefetcher global first, add an assertion that the address table has exactly `n_layers * n_tensors` entries before launching `ttnn.dram_prefetcher`, and add a warmup unit/static test that validates the batch-32 page table contains no mixed `[-1, 0, ...]` inactive rows.

## Continue In Codex

To reopen this investigation thread:

```bash
codex resume 019f1d1f-b98f-7fc1-b6c4-5fa3d382503e
```

</details>

**Comment by ifabijanic-tt** [comment]
Looking at this issue

**Comment by tchedaTT** [comment]
Looks like some runs are not crashing, but are instead outputting garbage tokens https://github.com/tenstorrent/tt-metal/actions/runs/28833171825/job/85512618198#step:17:3190
This strongly points towards bad memory writes as the cause

**Comment by ifabijanic-tt** [comment]
The issue was a part of larger sampling issues on Llama 70B on Galaxy machine. PRs for this issue are these (both on tt-metal and vllm side):
https://github.com/tenstorrent/tt-metal/pull/50688
https://github.com/tenstorrent/tt-metal/pull/50687
https://github.com/tenstorrent/tt-metal/pull/50685
https://github.com/tenstorrent/vllm/pull/444

All PRs are merged so this issue shouldn't happen as of today. Closing this issue. If anything new comes up, please open new issue.
